src/main/java/org/team1540/elmo/subsystems/drivetrain/DriveTrain.java

```diff
@@ -26,12 +31,165 @@ public DriveTrain() {
+    public void periodic() {
+        super.periodic();
+        SmartDashboard.putNumber(getName()+"/navx-angle", getAngle());
+
+        rampingPeriodic();
+    }
+
     public void stop() {
         setPercent(0, 0);
     }
+    public void stopRamped() {
+        setPercentRamped(0, 0);
+    }
 
-    public void setPercent(double leftPercent, double rightPercent) {
+    ///////////// RAMPING ////////////
+    private enum DriveRampingMode {
+        NO_RAMP,
+        SPEED_SPIN,
+        DIFFERENTIAL,
+    }
+    private DriveRampingMode driveRampingMode = DriveRampingMode.DIFFERENTIAL;
+
+    private void rampingPeriodic() {
+        if(driveRampingMode.equals(DriveRampingMode.SPEED_SPIN)) setSpeedSpinRaw(speedRateLimiter.calculate(speedRampedSetpoint),spinRateLimiter.calculate(spinRampedSetpoint));
+        else if(driveRampingMode.equals(DriveRampingMode.DIFFERENTIAL)) {
+            frontLeft.set(ControlMode.PercentOutput, leftRateLimiter.calculate(leftRampedSetpoint));
+            frontRight.set(ControlMode.PercentOutput, rightRateLimiter.calculate(rightRampedSetpoint));
+        }
+    }
+
+    ///////////////// Differential Drive Mode /////////////////
+    private void setPercentRaw(double leftPercent, double rightPercent) {
         frontLeft.set(ControlMode.PercentOutput, leftPercent);
         frontRight.set(ControlMode.PercentOutput, rightPercent);
     }
+    public double getLeftMotorPercent() { return frontLeft.getMotorOutputPercent(); }
+    public double getRightMotorPercent() { return frontRight.getMotorOutputPercent(); }
+
+    public void setPercent(double leftPercent, double rightPercent) {
+        // set
+        setPercentRaw(leftPercent, rightPercent);
+
+        // ramping
+        driveRampingMode = DriveRampingMode.NO_RAMP;
+        leftRateLimiter.reset(leftPercent);
+        rightRateLimiter.reset(rightPercent);
+    }
+
+    ///////////////// Differential Drive Mode Ramping /////////////////
+    private final double defaultAccLimit = 5;
+    private final double defaultDecLimit = 3;
+    private double leftRampedSetpoint = 0;
+    private double rightRampedSetpoint = 0;
+    private SignedSlewRateLimiter leftRateLimiter = new SignedSlewRateLimiter(
+            SmartDashboard.getNumber(getName() + "/AccelerateLimit", defaultAccLimit),
+            SmartDashboard.getNumber(getName() + "/DecelerateLimit", defaultDecLimit)
+    );
+    private SignedSlewRateLimiter rightRateLimiter = new SignedSlewRateLimiter(
+            SmartDashboard.getNumber(getName() + "/AccelerateLimit", defaultAccLimit),
+            SmartDashboard.getNumber(getName() + "/DecelerateLimit", defaultDecLimit)
+    );
+
+    public void setPercentRamped(double leftPercent, double rightPercent) {
+        // if setting right after using speed-spin ramped mode, translate units
+        if(driveRampingMode.equals(DriveRampingMode.SPEED_SPIN)) {
+            double left = speedRateLimiter.getPrevVal() + spinRateLimiter.getPrevVal()/2;
```

Comment:
Why 1.5x the previous value?

Reply:
I think you misread- its **speed** + spin/2 instead of spin + spin/2. I did some algebra to figure out how to translate between specifications, it might be wrong, but i believe its right..?

Reply:
An, you're correct in my misreading. I'm not going to try and decipher all the math you did for this, I just want to make sure it doesn't become a time sink